ext/go-plugin/vagrant/plugin/capabilities.go

```diff
@@ -0,0 +1,396 @@
+package plugin
+
+import (
+	"context"
+	"encoding/json"
+
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+
+	go_plugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vagrant/ext/go-plugin/vagrant"
+	"github.com/hashicorp/vagrant/ext/go-plugin/vagrant/plugin/proto"
+
+	"github.com/LK4D4/joincontext"
+)
+
+type GuestCapabilities interface {
+	vagrant.GuestCapabilities
+	Meta
+}
+
+type GuestCapabilitiesPlugin struct {
+	go_plugin.NetRPCUnsupportedPlugin
+	Impl GuestCapabilities
+}
+
+func (g *GuestCapabilitiesPlugin) GRPCServer(broker *go_plugin.GRPCBroker, s *grpc.Server) error {
+	g.Impl.Init()
+	vagrant_proto.RegisterGuestCapabilitiesServer(s, &GRPCGuestCapabilitiesServer{
+		Impl: g.Impl,
+		GRPCIOServer: GRPCIOServer{
+			Impl: g.Impl}})
+	return nil
+}
+
+func (g *GuestCapabilitiesPlugin) GRPCClient(ctx context.Context, broker *go_plugin.GRPCBroker, c *grpc.ClientConn) (interface{}, error) {
+	client := vagrant_proto.NewGuestCapabilitiesClient(c)
+	return &GRPCGuestCapabilitiesClient{
+		client:  client,
+		doneCtx: ctx,
+		GRPCIOClient: GRPCIOClient{
+			client:  client,
+			doneCtx: ctx}}, nil
+}
+
+type GRPCGuestCapabilitiesServer struct {
+	GRPCIOServer
+	Impl GuestCapabilities
+}
+
+func (s *GRPCGuestCapabilitiesServer) GuestCapabilities(ctx context.Context, req *vagrant_proto.Empty) (resp *vagrant_proto.SystemCapabilityList, err error) {
+	resp = &vagrant_proto.SystemCapabilityList{}
+	g, _ := errgroup.WithContext(ctx)
+	g.Go(func() (err error) {
+		r, err := s.Impl.GuestCapabilities()
+		if err != nil {
+			return
+		}
+		for _, cap := range r {
+			rcap := &vagrant_proto.SystemCapability{Name: cap.Name, Platform: cap.Platform}
+			resp.Capabilities = append(resp.Capabilities, rcap)
+		}
+		return
+	})
+	err = g.Wait()
```

Comment:
Nice edits !! I also would like to warn you here that this wait will block until the closure returns.

The ctx in `g, thisContext := errgroup.WithContext(ctx)` will be cancelled at the moment the closure returns or the top context is returned.

```suggestion
	<-ctx.Done()
	return ctx.Error() // this will ignore the error in the closure though
```